File: FBDeviceControl/Management/FBDeviceManager.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceManager.h"

#import "FBAMDevice+Private.h"
#import "FBDeviceControlError.h"
#import "FBDeviceControlFrameworkLoader.h"
#import "FBDeviceStorage.h"

@implementation FBDeviceManager

@synthesize delegate = _delegate;

- (instancetype)initWithLogger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _logger = logger;
  _storage = [[FBDeviceStorage alloc] initWithLogger:logger];

  return self;
}

- (void)dealloc
{
  [self stopListeningWithError:nil];
}

#pragma mark Implemented in Subclasses

- (BOOL)startListeningWithError:(NSError **)error
{
  return [[FBDeviceControlError
    describeFormat:@"-[%@ %@] is abstract and should be overridden", NSStringFromClass(self.class), NSStringFromSelector(_cmd)]
    failBool:error];
}

- (BOOL)stopListeningWithError:(NSError **)error
{
  return [[FBDeviceControlError
    describeFormat:@"-[%@ %@] is abstract and should be overridden", NSStringFromClass(self.class), NSStringFromSelector(_cmd)]
    failBool:error];
}

- (id)constructPublic:(PrivateDevice)privateDevice identifier:(NSString *)identifier info:(NSDictionary<NSString *,id> *)info
{
  NSAssert(NO, @"-[%@ %@] is abstract and should be overridden", NSStringFromClass(self.class), NSStringFromSelector(_cmd));
  return nil;
}

+ (void)updatePublicReference:(id)publicDevice privateDevice:(PrivateDevice)privateDevice identifier:(NSString *)identifier info:(NSDictionary<NSString *,id> *)info
{
  NSAssert(NO, @"-[%@ %@] is abstract and should be overridden", NSStringFromClass(self), NSStringFromSelector(_cmd));
}

+ (PrivateDevice)extractPrivateReference:(id)publicDevice
{
  NSAssert(NO, @"-[%@ %@] is abstract and should be overridden", NSStringFromClass(self), NSStringFromSelector(_cmd));
  return nil;
}

#pragma mark Called in Subclasses

- (void)deviceConnected:(PrivateDevice)privateDevice identifier:(NSString *)identifier info:(NSDictionary<NSString *,id> *)info
{
  [self.logger logFormat:@"Device Connected %@", privateDevice];

  // Make sure that we pull from all known instances created by this class.
  // We do this instead of the attached ones.
  // The reason for doing so is that consumers of these instances may be holding onto a reference to a device that's been re-connected.
  // Pulling from the map of referenced devices means that we re-use these referenced devices if they are present and the underlying reference is replaced.
  // If the device is no-longer referenced it will have been removed from the referencedDevices mapping as it's values are weakly-held.
  id device = [self.storage deviceForKey:identifier];
  if (device) {
    [self.logger.info logFormat:@"Device has been re-attached %@", device];
  } else {
    device = [self constructPublic:privateDevice identifier:identifier info:info];
    [self.logger.info logFormat:@"Created a new Device instance %@", device];
  }

  // See whether the Private API reference represents a replacement of something we already know bout.
  PrivateDevice oldPrivateDevice = [self.class extractPrivateReference:device];
  if (oldPrivateDevice == NULL) {
    [self.logger logFormat:@"New '%@' appeared for the first time", privateDevice];
    [self.class updatePublicReference:device privateDevice:privateDevice identifier:identifier info:info];
  } else if (privateDevice != oldPrivateDevice) {
    [self.logger logFormat:@"New '%@' replaces Old Device '%@'", privateDevice, oldPrivateDevice];
    [self.class updatePublicReference:device privateDevice:privateDevice identifier:identifier info:info];
  } else {
    [self.logger logFormat:@"Existing Device %@ is the same as the old", privateDevice];
  }

  // Update the internal state
  [self.storage deviceAttached:device forKey:identifier];

  // Notify the delegate.
  [self.delegate targetAdded:device inTargetSet:self];
}

- (void)deviceDisconnected:(PrivateDevice)privateDevice identifier:(NSString *)identifier
{
  [self.logger logFormat:@"Device Disconnected %@", privateDevice];
  id device = [self.storage deviceForKey:identifier];
  if (!device) {
    [self.logger logFormat:@"No Device named %@ from attached devices, nothing to remove", identifier];
    return;
  }
  [self.logger logFormat:@"Removing Device %@ from attached devices", identifier];

  // Update the internal state.
  [self.storage deviceDetachedForKey:identifier];

  // Notify the delegate
  [self.delegate targetRemoved:device inTargetSet:self];
}

#pragma mark Public

- (NSArray<id> *)currentDeviceList
{
  return [self.storage.attached.allValues sortedArrayUsingSelector:@selector(uniqueIdentifier)];
}

- (NSArray<id<FBiOSTargetInfo>> *)allTargetInfos
{
  return self.currentDeviceList;
}

- (NSString *)description
{
  return [NSString stringWithFormat:@"%@: %@", NSStringFromClass(self.class), [FBCollectionInformation oneLineDescriptionFromArray:self.allTargetInfos]];
}

@end

File: FBDeviceControl/Management/FBDeviceManager.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

/// Discovery of a set of devices, shared by the AMDevice and restorable device sources.
///
/// Subclasses supply how to listen and how to turn a private device reference into a public one;
/// this class owns the registry of what is currently attached and notifies the delegate.
class FBDeviceManager<PublicDevice: AnyObject>: NSObject, FBiOSTargetSet {

  // MARK: - Properties

  let logger: any FBControlCoreLogger
  let storage: FBDeviceStorage<PublicDevice>
  weak var delegate: (any FBiOSTargetSetDelegate)?

  // MARK: - Initializers

  init(logger: any FBControlCoreLogger) {
    self.logger = logger
    self.storage = FBDeviceStorage(logger: logger)
    super.init()
  }

  deinit {
    try? stopListening()
  }

  // MARK: - Implemented in Subclasses

  /// Starts listening for device notifications.
  func startListening() throws {
    throw FBDeviceManagerError.abstractMethod(name: "startListening")
  }

  /// Stops listening for device notifications.
  func stopListening() throws {
    throw FBDeviceManagerError.abstractMethod(name: "stopListening")
  }

  /// Constructs the type from the private one.
  func constructPublic(_ privateDevice: CFTypeRef, identifier: String, info: [String: Any]?) -> PublicDevice {
    fatalError("constructPublic is abstract and must be overridden")
  }

  /// Updates the type with data from the private one.
  class func updatePublicReference(
    _ publicDevice: PublicDevice,
    privateDevice: CFTypeRef,
    identifier: String,
    info: [String: Any]?
  ) {
    fatalError("updatePublicReference is abstract and must be overridden")
  }

  /// Extracts the private reference from the public one.
  class func extractPrivateReference(_ publicDevice: PublicDevice) -> Unmanaged<AnyObject>? {
    fatalError("extractPrivateReference is abstract and must be overridden")
  }

  // MARK: - Called in Subclasses

  func deviceConnected(_ privateDevice: CFTypeRef, identifier: String, info: [String: Any]?) {
    let privateAddress = Unmanaged.passUnretained(privateDevice as AnyObject).toOpaque()
    logger.log("Device Connected \(identifier) (\(privateAddress))")

    // Pull from all known instances created by this class rather than the attached ones: consumers
    // may be holding a device that has been re-connected, and re-using the referenced instance
    // means the underlying reference is replaced beneath them. A device that is no longer
    // referenced has already left the mapping, whose values are weakly held.
    let device: PublicDevice
    if let existing = storage.device(forKey: identifier) {
      logger.info().log("Device has been re-attached \(existing)")
      device = existing
    } else {
      device = constructPublic(privateDevice, identifier: identifier, info: info)
      logger.info().log("Created a new Device instance \(device)")
    }

    let oldPrivateDevice = Self.extractPrivateReference(device)
    if let oldPrivateDevice {
      if oldPrivateDevice.toOpaque() != privateAddress {
        logger.log("New '\(identifier)' (\(privateAddress)) replaces Old Device (\(oldPrivateDevice.toOpaque()))")
        Self.updatePublicReference(device, privateDevice: privateDevice, identifier: identifier, info: info)
      } else {
        logger.log("Existing Device '\(identifier)' (\(privateAddress)) is the same as the old")
      }
    } else {
      logger.log("New '\(identifier)' (\(privateAddress)) appeared for the first time")
      Self.updatePublicReference(device, privateDevice: privateDevice, identifier: identifier, info: info)
    }

    storage.deviceAttached(device, forKey: identifier)

    if let info = device as? any FBiOSTargetInfo {
      delegate?.targetAdded(info, in: self)
    }
  }

  func deviceDisconnected(_ privateDevice: CFTypeRef, identifier: String) {
    // The private ref may already be dead by the time a disconnect callback fires, so log it by
    // identifier and address only; never dereference it.
    let privateAddress = Unmanaged.passUnretained(privateDevice as AnyObject).toOpaque()
    logger.log("Device Disconnected \(identifier) (\(privateAddress))")
    guard let device = storage.device(forKey: identifier) else {
      logger.log("No Device named \(identifier) from attached devices, nothing to remove")
      return
    }
    logger.log("Removing Device \(identifier) from attached devices")

    storage.deviceDetached(forKey: identifier)

    if let info = device as? any FBiOSTargetInfo {
      delegate?.targetRemoved(info, in: self)
    }
  }

  // MARK: - Public

  var currentDeviceList: [PublicDevice] {
    Array(storage.attached.values).sorted { lhs, rhs in
      let lhsID = (lhs as? any FBiOSTargetInfo)?.uniqueIdentifier ?? ""
      let rhsID = (rhs as? any FBiOSTargetInfo)?.uniqueIdentifier ?? ""
      return lhsID < rhsID
    }
  }

  // MARK: - FBiOSTargetSet

  var allTargetInfos: [any FBiOSTargetInfo] {
    currentDeviceList.compactMap { $0 as? any FBiOSTargetInfo }
  }

  func target(withUDID udid: String) -> (any FBiOSTargetInfo)? {
    allTargetInfos.first { FBiOSTargetPredicateForUDID(udid).evaluate(with: $0) }
  }

  // MARK: - NSObject

  override var description: String {
    "\(type(of: self)): \(FBCollectionInformation.oneLineDescription(from: allTargetInfos))"
  }
}

public enum FBDeviceManagerError: Error, LocalizedError {
  case abstractMethod(name: String)

  public var errorDescription: String? {
    switch self {
    case let .abstractMethod(name):
      return "\(name) is abstract and must be overridden"
    }
  }
}
